[docs-logging]: move developer/architecture docs to user docs (#119125)
elastic/kibana: docs/setup/configuring-logging.asciidoc

[[logging-configuration]]
== Configure logging

The {kib} logging system has three main components: _loggers_, _appenders_ and _layouts_. These components allow us to log messages according to message type and level, to control how these messages are formatted and where the final logs will be displayed or stored.

- <<loggers-appenders-layout>>
- <<log-level>>
- <<logging-layouts>>
- <<logger-hierarchy>>

[float]
[[loggers-appenders-layout]]
=== Loggers, Appenders and Layouts
__Loggers__ define what logging settings should be applied to a particular logger.

__<<logging-appenders,Appenders>>__ define where log messages are displayed (eg. stdout or console) and stored (eg. file on the disk).

__<<logging-layouts,Layouts>>__ define how log messages are formatted and what type of information they include.

[float]
[[log-level]]
=== Log level

Currently we support the following log levels: _all_, _fatal_, _error_, _warn_, _info_, _debug_, _trace_, _off_.

Levels are ordered, so _all_ > _fatal_ > _error_ > _warn_ > _info_ > _debug_ > _trace_ > _off_.

> A log record is being logged by the logger if its level is higher than or equal to the level of its logger. Otherwise, the log record is ignored.

Reviewer: ```suggestion
A log record will be logged by the logger if its level is higher than or equal to the level of its logger. Otherwise, the log record is ignored.
```

> The _all_ and _off_ levels can only be used in configuration and are handy shortcuts that allow you to log every log record or disable logging entirely or for a specific logger. These levels can also be specified using <<logging-cli-migration,cli arguments>>.

Reviewer: ```suggestion
The _all_ and _off_ levels can only be used in configuration and are handy shortcuts that allow you to log every log record or disable logging entirely for a specific logger. These levels can also be specified using <<logging-cli-migration,cli arguments>>.
```

| `%date{ISO8601_TZ}`
| `2012-02-01T09:30:22.011-05:00`   `ISO8601` with timezone

| `%date{ISO8601_TZ}{America/Los_Angeles}`
| `2012-02-01T06:30:22.011-08:00`

| `%date{ABSOLUTE}`
| `09:30:22.011`

| `%date{ABSOLUTE}{America/Los_Angeles}`
| `06:30:22.011`

| `%date{UNIX}`
| `1328106622`

| `%date{UNIX_MILLIS}`
| `1328106622011`

|===

**pid**
Outputs the process ID.

The pattern layout also offers a `highlight` option that allows you to highlight
some parts of the log message with different colors. Highlighting is quite handy if log messages are forwarded
to a terminal with color support.

[float]
[[json-layout]]
==== JSON layout
With `json` layout log messages will be formatted as JSON strings in https://www.elastic.co/guide/en/ecs/current/ecs-reference.html[ECS format] that includes a timestamp, log level, logger, message text and any other metadata that may be associated with the log message itself.

[float]
[[logger-hierarchy]]
=== Logger hierarchy

Every logger has a unique name that follows a hierarchical naming rule. The logger is considered to be an
ancestor of another logger if its name followed by a `.` is a prefix of the descendant logger. For example, a logger

> named `a.b` is an ancestor of logger `a.b.c`. All top-level loggers are descendants of a special `root` logger at the top of the logger hierarchy. The `root` logger always exists, is fully configured and logs to `info` level be default. The `root` logger must also be configured if any other logging configuration is specified in your `kibana.yml`.

Reviewer: ```suggestion
named `a.b` is an ancestor of logger `a.b.c`. All top-level loggers are descendants of a special `root` logger at the top of the logger hierarchy. The `root` logger always exists, is fully configured and logs to `info` level by default. The `root` logger must also be configured if any other logging configuration is specified in your `kibana.yml`.
```

[[request-response-logger]]
**Request and Response Logs**

The `http.server.response` logger is configured with `debug` level and will automatically output
data about http requests and responses occurring on the {kib} server.
The message contains some high-level information, and the corresponding log meta contains the following:

[options="header"]
|===

| Meta property | Description | Format

| client.ip | IP address of the requesting client | ip

| http.request.method | http verb for the request (uppercase) | string

| http.request.mime_type | (optional) mime as specified in the headers | string

| http.request.referrer | (optional) referrer | string

| http.request.headers | request headers | object

| http.response.body.bytes | (optional) Calculated response payload size in bytes | number

| http.response.status_code | status code returned | number

| http.response.headers | response headers | object

| http.response.responseTime | (optional) Calculated response time in ms | number

| url.path | request path | string

| url.query | (optional) request query string | string

| user_agent.original | raw user-agent string provided in request headers | string

|===

> [[logging-appenders]]

Author: ```suggestion
[float]
[[logging-appenders]]
```

[[logging-configuration]]

> == Configure logging

Author: This page serves as a more detailed guide about the logging system and goes into the details we used to have in the dev docs section.

**Request and Response Logs**

The `http.server.response` logger is configured with `debug` level and will automatically output
data about http requests and responses occurring on the {kib} server.
The message contains some high-level information, and the corresponding log meta contains the following:

[options="header"]
|===

| Meta property | Description | Format

| client.ip | IP address of the requesting client | ip

| http.request.method | http verb for the request (uppercase) | string

| http.request.mime_type | (optional) mime as specified in the headers | string

| http.request.referrer | (optional) referrer | string

| http.request.headers | request headers | object

| http.response.body.bytes | (optional) Calculated response payload size in bytes | number

| http.response.status_code | status code returned | number

| http.response.headers | response headers | object

| http.response.responseTime | (optional) Calculated response time in ms | number

| url.path | request path | string

| url.query | (optional) request query string | string

| user_agent.original | raw user-agent string provided in request headers | string

|===

[[logging-appenders]]

> === Appenders

Author: Appenders renders as a new page in the logs under the Configure logging section:
<img width="547" alt="configure-logging-sections" src="https://user-images.githubusercontent.com/11909450/142695812-f34b7d3c-f69d-47f9-b49f-2d21fd45f3fe.png">

Reviewer: I think we should add `[float]` above the section to avoid creating a new page.

Author: I wanted to avoid adding too much content on one page but could go either way.

      policy:
        type: meta
        mode: remove
        properties:
          - path: "http.request.headers" # remove all request headers
    censor-response-headers:
      type: rewrite
      appenders: [console] # output to console
      policy:
        type: meta
        mode: update
        properties:
          - path: "http.response.headers.set-cookie"
            value: "[REDACTED]"
----

[float]
====== Complete Example For Rewrite Appender

[source,yaml]
----
logging:
  appenders:
    custom_console:
      type: console
      layout:
        type: pattern
        highlight: true
        pattern: "[%date][%level][%logger] %message %meta"
    file:
      type: file
      fileName: ./kibana.log
      layout:
        type: json
    censor:
      type: rewrite
      appenders: [custom_console, file]
      policy:
        type: meta
        mode: update
        properties:
          - path: "http.request.headers.cookie"
            value: "[REDACTED]"
  loggers:
    - name: http.server.response
      appenders: [censor] # pass these logs to our rewrite appender
      level: debug
----

// EXAMPLES COPIED OVER FROM logging-settings

> [[log-settings-examples]]

Author: New subsection

[source,yaml]
----
logging:
  appenders:
    console:
      type: console
      layout:
        type: pattern
        highlight: true
    file:
      type: file
      fileName: /var/log/kibana.log
    custom:
      type: console
      layout:
        type: pattern
        pattern: "[%date][%level] %message"
    json-file-appender:
      type: file
      fileName: /var/log/kibana-json.log
      layout:
        type: json

  root:
    appenders: [console, file]
    level: error

  loggers:
    - name: plugins
      appenders: [custom]
      level: warn
    - name: plugins.myPlugin
      level: info
    - name: server
      level: fatal
    - name: optimize
      appenders: [console]
    - name: telemetry
      appenders: [json-file-appender]
      level: all
    - name: metrics.ops
      appenders: [console]
      level: debug
----

Here is what we get with the config above:
[options="header"]
|===

| Context name | Appenders | Level

| root | console, file | error

| plugins | custom | warn

| plugins.myPlugin | custom | info

| server | console, file | fatal

| optimize | console | error

| telemetry | json-file-appender | all

| metrics.ops | console | debug
|===

> === Cli configuration

Author: new subsection:
<img width="1261" alt="logging-cli-options" src="https://user-images.githubusercontent.com/11909450/142696216-9d498dc4-3ea5-4005-bed8-95abc2220a24.png">
